components/base/kinematicbase/kinematics.go

```diff
@@ -102,56 +107,106 @@ func (ddk *differentialDriveKinematics) GoToInputs(ctx context.Context, desired
 	if err != nil {
 		return err
 	}
-	validRegion, err := ddk.newValidRegionCapsule(current, desired)
+	distErr, headingErr, err := ddk.errorState(current, desired)
+	if err != nil {
+		return err
+	}
+	validRegion, err := ddk.newValidRegionCapsule(current, desired, distErr)
 	if err != nil {
 		return err
 	}
 
+	t := time.NewTimer(timeout)
+	poseChange := make(chan bool)
+	movementErr := make(chan error)
+
 	// this loop polls the error state and issues a corresponding command to move the base to the objective
 	// when the base is within the positional threshold of the goal, exit the loop
-	for err = ctx.Err(); err == nil; err = ctx.Err() {
-		col, err := validRegion.CollidesWith(spatialmath.NewPoint(r3.Vector{X: current[0].Value, Y: current[1].Value}, ""))
-		if err != nil {
-			return err
-		}
-		if !col {
-			return errors.New("base has deviated too far from path")
-		}
+	utils.PanicCapturingGo(func() {
+		prevDistErr := -1
+		prevHeadingErr := math.Inf(-1)
+		var contextErr error
+		for contextErr = ctx.Err(); contextErr == nil; contextErr = ctx.Err() {
+			if poseChanged(prevDistErr, prevHeadingErr, distErr, headingErr) {
+				poseChange <- true
```

Comment:
What you have here is a bit more complex than it needs to be, and may be prone to lockups if something goes wrong.

I would recommend removing the `poseChange` channel, the `time.NewTimer(timeout)`, and the entirety of the other goroutine which watches the `poseChange` channel.

All of that removed code can become just the following:
```
lastUpdate := time.Now()
for contextErr = ctx.Err(); contextErr == nil; contextErr = ctx.Err() {
	if poseChanged(prevDistErr, prevHeadingErr, distErr, headingErr) {
		lastUpdate = time.Now()
	} else if time.Since(lastUpdate) > timeout {
		throw an error
	}
```

Reply:
Should I keep the goroutine that the whole for loop exists in? I created that so it could run concurrently with the other goroutine which was watching for a pose change but it seems like it's no longer necessary?

Reply:
Good point, there's probably not a reason to use goroutines at all in that case.

Reply:
talked to @raybjork - looks like the `issueCommand` calls will block until they can successfully complete and this new structure won't be able to interrupt that, which is why the separate goroutine is necessary for the timer

Reply:
Got it. That makes sense.

In that case the correct structure is still not what we have here, but rather what we discussed above with no goroutines, but with the addition of a timeout to the context `context.WithTimeout(ctx, timeout)` each time we have a position update.